### DISCUSSION TITLE
I used the -l option to install, why is it not installed to the specified location on the hard disk?

### DISCUSSION THREAD

**Discussion by dong-lufei** [microsoft/winget-cli#2812] [Q&A] [answered]
winget install DuongDieuPhap.ImageGlass -l "D:\ImageGlass"
or
winget install ImageGlass -l "D:\ImageGlass"

Neither installed to location "D:\ImageGlass"，Instead it installed to the default location C:\Program Files\ImageGlass

**Comment by Trenly** [comment, marked as answer]
Not all installers support the `-l` switch. In some cases the installers can be marked as not having a location switch at all, and in other cases the location switch is simply unknown. 

https://github.com/microsoft/winget-cli/blob/fe1eef3400b6dc03df66a93211b85fec0d7a44e3/src/AppInstallerCLIPackage/Shared/Strings/en-us/winget.resw#L340-L342

**Reply by ccnly** [reply]
Ideally, winget would throw a warning if the installer doesn't recognize or use the `-l` switch.

**Reply by Trenly** [reply]
> Ideally, winget would throw a warning if the installer doesn't recognize or use the `-l` switch.

We do have the ability to add that information into the manifests if an app is known to truly not support a location switch. However, because it is an optional parameter, not all manifests contain that field and in many cases the installers do support a location switch, it just isn’t known (or hasn’t been added to the manifest).

**Reply by github-account1111** [reply]
> and in many cases the installers do support a location switch

What's the location switch? Is it `/l`? `/loc`? `-loc`? `--loc`? `--dest`? When I install a package with `winget install --location C:\install\path`, does winget just try a bunch of possible switches to see what sticks? Can't it throw a warning if nothing sticks?

**Reply by denelon** [reply]
Run the command with `--verbose-logs --logs` you will see all arguments used to execute the installer by WinGet. It may depend on the installer type and in many cases, WinGet has been built with the argument for the given installer type.

**Comment by dong-lufei** [comment]
Are there any configurations that I can download and install to the path "D:\winget" by default?

I install a software, such as winget install abcd, and then the software abcd is installed in the "d:\winget" folder.

**Comment by Trenly** [comment]
As of right now there is only the [defaultInstallRoot ](https://github.com/microsoft/winget-cli/blob/master/doc/Settings.md#default-install-root), but that is only activated if the installer requires a location

**Comment by michael-hiegl** [comment]
In my opinion, it's not ok when i'm using the --location flag and the application gets installed somewhere else regardless. Either make it mandatory for publishers that the --location flag takes effect or return a warning/info to the terminal which should be followed by a continue (Y/N) selection.

**Reply by denelon** [reply]
Some installers appear to support the argument, but in reality, they don't, or if the location isn't properly escaped (which depends on the shell) it may not work.

**Reply by github-account1111** [reply]
Can't winget abort if a package happens to not support the argument? Much better to not do enough than do too much. Also, what's the argument? `--path`?

**Reply by denelon** [reply]
In some cases, there is no way for WinGet to know if the argument for the location doesn't behave correctly. In those cases, the installer would need to be modified.
